### PR TITLE
WIP feat: Material baking

### PR DESCRIPTION
**Pull request by schinkowski**

### CODE REVIEW

File: Plugins~/Src/MeshSyncClientBlender/python/3.1.0/unity_mesh_sync.py [hunk 1]
```diff
@@ -340,26 +339,29 @@ def draw(self, context):
...
-)
+    MESHSYNC_Preferences
+] + sharedClasses
 
 def register():
-    msb_initialize_properties()
     bpy.app.handlers.load_post.append(MESHSYNC_OT_AutoSync.load_handler)
     for c in classes:
         bpy.utils.register_class(c)
+    msb_initialize_properties()
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/3.1.0/unity_mesh_sync.py:353]: why is this necessary? was it causing any issues because of the order?

File: Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py [hunk 2]
```diff
@@ -26,12 +27,6 @@
 mesh_sync_on_post_export = []
 
 
-class MESHSYNC_PT:
-    bl_space_type = "VIEW_3D"
-    bl_region_type = "UI"
-    bl_category = "Tool"
-
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py]: do we really need to remove this?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
 
 from .unity_mesh_sync_installation import *
 from . import MeshSyncClientBlender as ms
-from bpy.types import AddonPreferences
-from bpy.props import StringProperty, IntProperty, BoolProperty
+from bpy.types import AddonPreferences, PropertyGroup
+from bpy.props import StringProperty, IntProperty, BoolProperty, IntVectorProperty, CollectionProperty
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py:9]: why do we import these new classes?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+    bl_description = "Bakes textures, creates material copies and assigns the baked materials for all materials that " \
+                     "cannot be exported without baking them to textures"
+
+    def isMaterialCopy(self, mat):
+        return ORIGINAL_MATERIAL in mat
+
+    def canMaterialBeBaked(self, mat):
+        return mat is not None and\
+            not self.isMaterialCopy(mat) and\
+            mat.use_nodes
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:107]: why do we not bake copies?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+
+    def findMaterialOutputNode(self, mat):
+        node_tree = mat.node_tree
+
+        outputNode = None
+        for node in node_tree.nodes:
+            if node.type == 'OUTPUT_MATERIAL' and len(node.inputs[0].links) == 1:
+                outputNode = node
+                # Blender uses the last OUTPUT_MATERIAL node, so don't stop search here
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:116]: What is the order of output nodes when traversing the tree? is it guaranteed to be as we expect it, or it could change in the future?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+
+                bakedBSDF.location = (bsdf.location[0], minYLocation - 1000)
+                # Give bsdf a name and set its name on the material, so we can find it again:
+                bakedBSDF.name = BAKED_MATERIAL_SHADER
+                matCopy[BAKED_MATERIAL_SHADER] = bakedBSDF.name
+
+                matIndex = obj.material_slots.find(mat.name)
+                obj.material_slots[matIndex].material = matCopy
+            # obj.data.materials.pop(index=obj.data.materials.find(mat.name))
+            # obj.data.materials.append(matCopy)
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:436]: do we need anything from this dead code? could we remove it?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+        self.ensureUVs(context, obj)
+
+        scene = context.scene
+
+        if obj.mode == 'EDIT':
+            bpy.ops.object.mode_set(mode='OBJECT')
+
+        scene.render.engine = "CYCLES"
+        scene.cycles.device = "GPU"
+        scene.cycles.samples = 10       # TODO: This could be a setting for the user to change
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:449]: I agree with the TODO comment. Also, does cycles support all bsdf that Eevee supports?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+        scene.view_settings.view_transform = 'Standard'
+        scene.view_settings.look = 'None'
+        scene.view_settings.exposure = 0
+        scene.view_settings.gamma = 1
+        scene.sequencer_colorspace_settings.name = 'sRGB'
+        scene.cycles.time_limit = 0
+
+        # Set Cycles to use CUDA, if we aren't using any acceleration
+        if context.preferences.addons["cycles"].preferences.compute_device_type == "None":
+            context.preferences.addons["cycles"].preferences.compute_device_type = "CUDA"
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:466]: Is this the only option available? What about OpenCL?

Comment (seandillon92) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:466]: Also, would the Cuda option be ignored in non-Nvidia cards or will it throw an error?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+        scene.cycles.use_auto_tile = True
+        scene.cycles.max_bounces = 12
+        scene.cycles.diffuse_bounces = 8
+        scene.cycles.glossy_bounces = 2
+        scene.cycles.transparent_max_bounces = 0
+        scene.cycles.transmission_bounces = 2
+        scene.cycles.volume_bounces = 0
```
Comment (seandillon92) [inline Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:473]: Would it make sense to pass these values from the cycles panel?

**Review by seandillon92 — COMMENTED** [review]
Looks good overall

File: Plugins~/Src/MeshSyncClientBlender/python/3.1.0/unity_mesh_sync.py (hunk 1 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/3.1.0/unity_mesh_sync.py:353]: We are now using a property group and a collection properties of the new group, which requires the class for it to be registered before it can be used.

File: Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py (hunk 2 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/3.3.0/unity_mesh_sync.py]: I've just moved it to another file so it can be shared across different parts of the code more easily. This is the same for all versions and doesn't need a copy in each __init__ file.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py (hunk 3 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py:9]: I'm using them for baking

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 4 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:107]: This is to catch a scenario where a material copy we created is baked, which should never happen. That would bake textures that were already baked.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 5 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:116]: From my testing this is what blender does. The node order in the nodetree can change but blender always uses the last material output node. It might change but there is no other way for us to detect which material output node is used, it's undefined behaviour and up to the user to set up correctly.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 7 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:449]: As far as I know, cycles supports everything eevee supports and more.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 9 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:473]: We're changing the settings on the cycles panel by setting those here.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 8 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:466]: Need to check, maybe we can remove this entirely and let the user choose it.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py (hunk 3 above)
Comment (seandillon92) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py:9]: But why are they imported in preferences? could they not be imported in the baking module?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 4 above)
Comment (seandillon92) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:107]: Just to make sure I follow:
I create a material in Blender.
I want to create a variation of the material so I create a copy of the first.
I modify the copy.
Will my copy be baked?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 9 above)
Comment (seandillon92) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:473]: oh right, is this necessary though? Maybe its better to let the user configure these?

Comment (seandillon92) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:473]: also, if we decide to modify cycles settings through the script, should we revert them after the bake?

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py (hunk 4 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_baking.py:107]: If the copy was of an original, yes. If you copy a baked material, no. Modifications to baked materials cannot be supported unless we want to force the user to manually revert to the original material and just bake whatever is on the objects.

File: Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py (hunk 3 above)
Comment (schinkowski) [reply Plugins~/Src/MeshSyncClientBlender/python/unity_mesh_sync_preferences.py:9]: They could, but all the property imports used by other modules are in this file so I wanted to keep that consistent.
